### PR TITLE
Add Respawn Test and ClientSim Method

### PR DESCRIPTION
**Pull request by cubedparadox**
Added a world test to reproduce the respawn issue in clientsim, and added the missing Respawn() method to the ClientSimPlayerManager to resolve the issue.

### CODE REVIEW

**Comment by CyanLaser** [comment]
Thank you for pinging me and letting me know. Yes I would be interested in the future as well.

File: Runtime/System/ClientSimPlayerManager.cs [hunk 1]
```diff
@@ -322,6 +322,12 @@ public static void TeleportTo(VRCPlayerApi player, Vector3 position, Quaternion
         {
             TeleportToOrientationLerp(player, position, rotation, VRC_SceneDescriptor.SpawnOrientation.Default, false);
         }
+        
+        
+        public static void Respawn(VRCPlayerApi playerApi)
+        {
+            playerApi.GetPlayerController().Respawn();
```
Comment (CyanLaser) [inline Runtime/System/ClientSimPlayerManager.cs:329]: I believe this will crash if the player isn't local as GetPlayerController() will return null.
https://github.com/vrchat-community/ClientSim/blob/main/Runtime/Utilities/ClientSimExtensions.cs#L10-L23

While this solution will work and is probably the minimum amount of new coded added, I wonder about directly exposing the Respawn method. Convention in ClientSim seems a little mixed between direct exposing and using the EventDispatcher to send and handle different events. It is up to you on which version you should use. 

If you go the event route, I recommend moving this to ClientSimMain and creating a non-static method that dispatches the event. Here are some references to the Respawn event used in the menu as an example:

Respawn from menu event:
https://github.com/vrchat-community/ClientSim/blob/main/Runtime/Events/EventTypes/ClientSimMenuEvents.cs#L8

Menu sending the respawn event:
https://github.com/vrchat-community/ClientSim/blob/main/Runtime/System/ClientSimMenu.cs#L415-L420

Player controller's handling of the menu respawn event:
https://github.com/vrchat-community/ClientSim/blob/main/Runtime/Player/ClientSimPlayerController.cs#L154-L157

The player controller subscribes to events here:
https://github.com/vrchat-community/ClientSim/blob/main/Runtime/Player/ClientSimPlayerController.cs#L99-L122

Player specific events are located in this file and a generic player respawn should be put here:
https://github.com/vrchat-community/ClientSim/blob/main/Runtime/Events/EventTypes/ClientSimOnPlayerEvents.cs

File: Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            // Walk into the respawn cube.
+            Helper.TestInput.SetInputRun(true);
+
+            yield return Helper.WalkToPoint(testHelpers.respawnCube.gameObject.transform, "Didn't reach RespawnCube");
+
+            // Wait for the same amount of time the UdonBehaviour uses to delay the Respawn call
+            yield return new WaitForSeconds(testHelpers.respawnCube.GetProgramVariable<float>("delay"));
+
+            //Verify that the player is very close to Spawn
+            Vector3 playerPosition = Helper.GetLastJoinPlayer().GetPosition();
```
Comment (CyanLaser) [inline Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs:82]: As a note, you can still use Networking.LocalPlayer here if you want to explicitly get the player you are controlling.

File: Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        private static readonly string _sceneName = "ClientSimIssue3RespawnTest";
+        private static readonly string _samplesDirectory = Path.Combine("Assets", "Samples", "VRChat Client Simulator");
+        private static bool _sceneExists;
+        
+        protected override ClientSimSettings GetTestSettings()
+        {
+            return new ClientSimSettings
+            {
+                enableClientSim = true,
+                initializationDelay = 0,
```
Comment (CyanLaser) [inline Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs:25]: I found that leaving this at zero will not catch Udon errors that run in start. If you add a small delay here, it will catch them. Obviously not needed for this test, but might be useful in the future. 0.5f seemed to work for me, but I'm not sure if this is machine dependent.

**Review by CyanLaser — COMMENTED** [review]
Thank you for implementing this API and adding tests to verify it! 
Also please let me know if I am overstepping by commenting too much on this PR.

Overall it seems fine and will solve the issue for the Respawn method. Could you please also add the RespawnWithIndex method as well?

I added a few nits in some areas that are up to you on how you want to handle, or not. 

Could you also add a new Integration test that verifies the API call directly and not through Udon? You can add to the end of the current Respawn test, checking everything the same way, but using the new API to respawn the player instead of the Helper method.
https://github.com/vrchat-community/ClientSim/blob/main/Tests/Runtime/IntegrationTests/ClientSimPlayerApiTests.cs#L527

File: Runtime/System/ClientSimMain.cs [hunk 4]
```diff
@@ -529,6 +530,7 @@ private void RemoveSDKLinks()
             VRCPlayerApi._SetGravityStrength -= ClientSimPlayerManager.SetGravityStrength;
             VRCPlayerApi.IsGrounded -= ClientSimPlayerManager.IsGrounded;
             VRCPlayerApi._UseLegacyLocomotion -= ClientSimPlayerManager.UseLegacyLocomotion;
+            VRCPlayerApi._Respawn += ClientSimPlayerManager.Respawn;
```
[the remark this answers is not in the document]
Comment (momo-the-monster) [reply Runtime/System/ClientSimMain.cs]: Fixed, thanks for catching.

File: Runtime/System/ClientSimMain.cs [hunk 5]
```diff
@@ -415,6 +415,7 @@ private void SetupSDKLinks()
             VRCPlayerApi._SetGravityStrength += ClientSimPlayerManager.SetGravityStrength;
             VRCPlayerApi.IsGrounded += ClientSimPlayerManager.IsGrounded;
             VRCPlayerApi._UseLegacyLocomotion += ClientSimPlayerManager.UseLegacyLocomotion;
+            VRCPlayerApi._Respawn += ClientSimPlayerManager.Respawn;
```
[the remark this answers is not in the document]
Comment (momo-the-monster) [reply Runtime/System/ClientSimMain.cs:418]: Added method and tests in the latest push.

File: package.json [hunk 6]
```diff
@@ -18,7 +18,12 @@
 			"displayName" : "ClientSimWorldTestExample",
 			"description" : "An example of a world test using the Test Framework included in ClientSim. Automate interactions in your world and verify the results. Currently requires knowledge on C# and Unity Testing.",
 			"path" : "Samples~/ClientSimWorldTestExample"
-		}
+		},
+        {
+          "displayName" : "ClientSimIssue3RespawnTest",
+          "description" : "A  world test to reproduce the ClientSim Respawn issue.",
+          "path" : "Samples~/ClientSimIssue3RespawnTest"
+        }
```
[the remark this answers is not in the document]
Comment (momo-the-monster) [reply package.json:26]: Agreed - we should either merge this test into the existing sample or at least batch them based on functionality. Tracking it in #10

File: Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            }
+        }
+
+        [SetUp]
+        public void CheckIfSceneExists()
+        {
+            Assert.IsTrue(_sceneExists, $"Failed to find Scene {_sceneName}! Please re-import the ClientSimWorldTestExample.");
+        }
+        
+        // Test will go through the example scene. This scene is a small "puzzle" requiring multiple steps to reach the end area.
```
[the remark this answers is not in the document]
Comment (momo-the-monster) [reply Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs]: Updated, thanks.

File: Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTestObjectReferences.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+namespace ClientSimTest
+{
+    // This class is only to make getting references to objects in the scene easier. 
+    // This can also be done without a dedicated MonoBehaviour and search the scene based on object names.
+    [AddComponentMenu("")]
+    public class ClientSimIssue3RespawnTestObjectReferences : MonoBehaviour
+    {
+        public UdonBehaviour respawnCube;
+        public Transform spawn;
+    }
```
[the remark this answers is not in the document]
Comment (momo-the-monster) [reply Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTestObjectReferences.cs]: I was thinking the same thing when getting / setting UdonBehaviour program variables - we could provide a place to list them, check they exist, etc. Made an item for this in #11

File: Runtime/System/ClientSimPlayerManager.cs (hunk 1 above)
Comment (momo-the-monster) [reply Runtime/System/ClientSimPlayerManager.cs:335]: Now checking for isLocal first, and logging a warning when trying to Respawn remote players.
Looking at the way you've implemented similar methods, I think this makes sense and matches the other code. If you have ideas / logic about when we should use one or the other, let me know so we can build the rules around it.

File: Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs (hunk 2 above)
Comment (momo-the-monster) [reply Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs]: Thanks, updated.

File: Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs (hunk 3 above)
Comment (momo-the-monster) [reply Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs]: Thanks, updated to 0.5 now and added a comment to this effect.

File: Runtime/Player/ClientSimPlayerController.cs [hunk 9]
```diff
@@ -200,12 +200,18 @@ public bool IsGrounded()
...
-        private void Respawn()
+        public void Respawn()
         {
             Teleport(_sceneManager.GetSpawnPoint(false), false);
             _eventDispatcher.SendEvent(new ClientSimOnPlayerRespawnEvent { player = _playerApi.Player });
         }
 
+        public void Respawn(int index)
+        {
+            Teleport(_sceneManager.GetSpawnPoint(index), false);
```
Comment (CyanLaser) [inline Runtime/Player/ClientSimPlayerController.cs:211]: How does VRChat handle different indices? I assume it will also throw an error and crash Udon? Might be worth adding a note of the comparison or that this method is allowed to crash.

File: Runtime/System/ClientSimPlayerManager.cs (hunk 1 above)
Comment (CyanLaser) [reply Runtime/System/ClientSimPlayerManager.cs:335]: This implementation is fine and matches other areas. If it was something else that didn't have an exposed way to get the object, I feel that case should instead use the EventDispatcher to prevent introducing new dependencies.

File: Runtime/System/ClientSimSceneManager.cs [hunk 10]
```diff
@@ -114,6 +114,22 @@ public Transform GetSpawnPoint(bool remote = false)
...
         }
+
+        public Transform GetSpawnPoint(int index)
+        {
+            if (!HasSceneDescriptor())
+            {
+                throw new ClientSimException("Cannot get spawn point when there is no scene descriptor.");
+            }
+            
+            if (_descriptor.spawns.Length <= index || _descriptor.spawns[index] == null)
```
Comment (CyanLaser) [inline Runtime/System/ClientSimSceneManager.cs:125]: Can you also check if index is greater than or equal to zero? If someone puts a negative number, the error will be an array out of bounds exception rather than a ClientSimException.

File: Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+                respawnEventFromDispatcherTriggered = true;
+            }
+            EventDispatcher.Subscribe<ClientSimOnPlayerRespawnEvent>(OnPlayerRespawn);
+
+            // Call Respawn static method on PlayerManager
+            var remotePlayer = Helper.SpawnRemotePlayer("RemotePlayer");
+            ClientSimPlayerManager.Respawn(remotePlayer);
+            
+            // Ensure OnPlayerRespawn fired by EventDispatcher
+            Assert.IsTrue(respawnEventFromDispatcherTriggered);
```
Comment (CyanLaser) [inline Samples~/ClientSimIssue3RespawnTest/ClientSimIssue3RespawnTest.cs:165]: If I'm reading this right, you are trying to respawn a remote player. Now that the api logs a warning for non local players, wouldn't this test fail as `respawnEventFromDispatcherTriggered` should still be false? It feels like a bug in ClientSim if a remote player can trigger the player respawned event. 

In some of the other tests, logs are verified when we expect the method to return early. It might be worth adding that here.
https://github.com/vrchat-community/ClientSim/blob/7b3a59c809730f8d47da31f3aeaa93e42e88786f/Tests/Runtime/IntegrationTests/ClientSimPlayerApiTests.cs#L453

File: Tests/Runtime/IntegrationTests/ClientSimPlayerApiTests.cs [hunk 12]
```diff
@@ -566,6 +566,19 @@ void OnPlayerRespawn(ClientSimOnPlayerRespawnEvent respawnEvent)
...
+            Assert.IsFalse(playerRespawned);
+            Networking.LocalPlayer.Respawn();
+            Assert.True(playerRespawned);
+            
+            // Test calling the Udon Respawn Method with an integer
+            playerRespawned = false;
+            Assert.IsFalse(playerRespawned);
+            var player = Networking.LocalPlayer;
+            player.Respawn(0);
+            Assert.True(playerRespawned);
```
Comment (CyanLaser) [inline Tests/Runtime/IntegrationTests/ClientSimPlayerApiTests.cs:580]: Thank you for adding these cases!

**Review by CyanLaser — APPROVED** [review]
Other than the question on the potentially failing test case, everything looks good with some optional nits.

File: Runtime/System/ClientSimSceneManager.cs (hunk 10 above)
Comment (momo-the-monster) [reply Runtime/System/ClientSimSceneManager.cs]: Yes - added this check to revert to 0 if the index is invalid.

**Comment by momo-the-monster** [comment]
@CyanLaser - the last change I pushed stops playback if you're not in a test and ClientSim initialization fails - just wanted to run this by you since it's a big change, but I think it makes it more clear what's going on - if something throws an Exception during initialization, the program should not continue IMO.

**Review by CyanLaser — APPROVED** [review]
This PR is good as it is, but one concern that could be addressed later.

Exiting playmode when there is an exception makes sense for the general use case. My only concern would be when someone is trying something unexpected with a VRChat project. For example, if there is no world descriptor because the scene open isn't a vrchat world, pressing playmode will now instantly exit instead of logging something but allowing the user to continue anyway. In that case, the user can open the settings to disable ClientSim, but just a little extra friction. Hopefully people can discover the settings menu under the VRChat/Utilities since that isn't as intuitive. 

It might be worth downgrading the exception to an error and return early, while verifying tests still behave as expected. Or just postponing this to another PR since this one seems to keep growing.

https://github.com/vrchat-community/ClientSim/blob/042856263bd6fc0835e50eabd6822a1dbc7a067c/Runtime/System/ClientSimMain.cs#L168-L173

File: Runtime/System/ClientSimMain.cs [hunk 13]
```diff
@@ -468,8 +480,12 @@ private void RemoveSDKLinks()
...
             VRCPlayerApi._UseAttachedStation -= ClientSimStationHelper.UseAttachedStation;
 
-            VRC_UiShape.GetEventCamera -= _player.GetCameraProvider().GetCamera;
-            
+            // Player can be invalid if we weren't able to spawn
+            if (_player)
+            {
+                VRC_UiShape.GetEventCamera -= _player.GetCameraProvider().GetCamera;
+            }
+
```
[the remark this answers is not in the document]
Comment (momo-the-monster) [reply Runtime/System/ClientSimMain.cs:497]: I added this specifically because _player is invalid at this point and was throwing an error in the console which was a distraction from the main issue (ClientSim failed to initialize, player failed to spawn). I don't understand what you mean by "added to the area adding the api", can you rephrase that?

File: Runtime/System/ClientSimTooltipManager.cs [hunk 14]
```diff
@@ -73,6 +73,10 @@ private void DisableTooltip(ClientSimTooltip tooltip)
 
         private void LateUpdate()
         {
+            if (_trackingProvider == null)
+            {
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (momo-the-monster) [reply Runtime/System/ClientSimTooltipManager.cs:79]: Possibly not - it was logging an error at some point in this process which was distracting from the main error, so I thought I'd just catch it here.

File: Runtime/System/ClientSimMain.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (CyanLaser) [reply Runtime/System/ClientSimMain.cs:497]: It's probably not an issue, but I was referring to this:
https://github.com/vrchat-community/ClientSim/blob/042856263bd6fc0835e50eabd6822a1dbc7a067c/Runtime/System/ClientSimMain.cs#L354

[the remark this answers is not in the document]
Comment (momo-the-monster) [reply Runtime/System/ClientSimMain.cs:497]: Thanks for the clarification. It's not specifically an issue in the place you mention because the enclosing function `SetupSDKLinks` doesn't run if ClientSim fails to initialize, but RemoveSDKLinks does. However, I think they should be matched so I'll push an update with a null check before this call as well.
